### PR TITLE
New gcs client

### PR DESCRIPTION
**Pull request by micafan**
## What do these changes do?

Add New GCS Client
@kfstorm  @raulchen @zhijunfu  @jovany-wang 

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: src/ray/gcs/actor_state_accessor.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  Status AsyncGet(const DriverID &driver_id, const ActorID &actor_id,
+                  DatumCallback<ActorTableDataT>::SingleItem callback);
+
+  Status AsyncGetCheckpointIds(
+      const DriverID &driver_id, const ActorID &actor_id,
+      DatumCallback<ActorCheckpointIdDataT>::SingleItem callback);
+
+ private:
+  AsyncGcsClient *client_impl_{nullptr};
```
Comment (zhijunfu) [inline src/ray/gcs/actor_state_accessor.h]: can we use smart pointers instead?  same for other places which use raw pointers

File: src/ray/gcs/actor_state_accessor.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace ray {
+
+namespace gcs {
+
+class AsyncGcsClient;
+
+class ActorStateAccessor {
+ public:
+  ActorStateAccessor(AsyncGcsClient *client_impl);
```
Comment (zhijunfu) [inline src/ray/gcs/actor_state_accessor.h]: can reference be used here?

File: src/ray/gcs/actor_state_accessor.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+  ~ActorStateAccessor() {}
+
+  Status AsyncGet(const DriverID &driver_id, const ActorID &actor_id,
+                  DatumCallback<ActorTableDataT>::SingleItem callback);
```
Comment (zhijunfu) [inline src/ray/gcs/actor_state_accessor.h]: `DatumCallback`  typo?

File: src/ray/gcs/actor_state_accessor.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+#include "ray/common/id.h"
+#include "ray/gcs/call_back.h"
+#include "ray/gcs/format/gcs_generated.h"
+
+namespace ray {
+
+namespace gcs {
+
+class AsyncGcsClient;
+
```
Comment (zhijunfu) [inline src/ray/gcs/actor_state_accessor.h]: Add some comments for this class? 

same for other classes

File: src/ray/gcs/gcs_client.cc [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  delete client_impl_;
+  client_impl_ = nullptr;
+
+  if (!thread_pool_.empty()) {
+    delete io_service_;
+    io_service_ = nullptr;
+  }
+}
+
+Status GcsClient::Connect() {
```
Comment (zhijunfu) [inline src/ray/gcs/gcs_client.cc]: Consider full initialization in constructor?

File: src/ray/gcs/gcs_client.cc [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+GcsClient::GcsClient(ClientOption option, ClientInfo info,
+                     boost::asio::io_service &io_service)
+    : option_(std::move(option_)), info_(std::move(info)), io_service_(&io_service) {}
+
+GcsClient::GcsClient(ClientOption option, ClientInfo info)
+    : option_(std::move(option_)), info_(std::move(info)) {}
+
+GcsClient::~GcsClient() {
+  delete node_state_accessor_;
```
Comment (zhijunfu) [inline src/ray/gcs/gcs_client.cc]: smart pointers?

File: src/ray/gcs/actor_state_accessor.h (hunk 4 above)
Comment (micafan) [reply src/ray/gcs/actor_state_accessor.h]: sure

File: src/ray/gcs/actor_state_accessor.h (hunk 2 above)
Comment (micafan) [reply src/ray/gcs/actor_state_accessor.h]: ref or pointer, same thing, right?

File: src/ray/gcs/gcs_client.h [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    RAY_CHECK(node_state_accessor_ != nullptr);
+    return *node_state_accessor_;
+  }
+
+  ActorStateAccessor &Actor() {
+    RAY_CHECK(actor_state_accessor_ != nullptr);
+    return *actor_state_accessor_;
+  }
+
+  TaskStateAccessor &Task() {
```
Comment (raulchen) [inline src/ray/gcs/gcs_client.h]: rename `Task` to `Tasks`? we already use this naming convention in core worker

File: src/ray/gcs/actor_state_accessor.cc [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+  const ClientID &client_id = client_impl_.GetClientInfo().id_;
+  ActorTable &actor_table = client_impl_.AsyncClient().actor_table();
+  return actor_table.Subscribe(driver_id, client_id, on_subscribe, on_done);
+}
+
+Status ActorStateAccessor::AsyncGetCheckpointIds(
+    const DriverID &driver_id, const ActorID &actor_id,
+    const DatumCallback<ActorCheckpointIdData>::OptionalItem &callback) {
+  auto on_successed = [callback](AsyncGcsClient *client, const ActorID &actor_id,
```
Comment (raulchen) [inline src/ray/gcs/actor_state_accessor.cc]: `on_success` or `on_succeeded`

File: src/ray/gcs/call_back.h [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+namespace gcs {
+
+using StatusCallback = std::function<void(Status status)>;
+
+template <typename T>
+class DatumCallback {
+ public:
+  using SingleItem = std::function<void(Status status, boost::optional<T> datum)>;
+
+  using MultiItem = std::function<void(Status status, std::vector<T> datums)>;
```
Comment (raulchen) [inline src/ray/gcs/call_back.h]: `datums` -> `data`

File: src/ray/gcs/call_back.h [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+};
```
Comment (raulchen) [inline src/ray/gcs/call_back.h]: These two callback types don't have to be defined in the same class? Just define 2 separate items.

File: src/ray/gcs/gcs_client_impl.cc [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+GcsClientImpl::~GcsClientImpl() {
+  if (!thread_pool_.empty()) {
+    RAY_CHECK(io_service_->stopped());
+    delete io_service_;
+    io_service_ = nullptr;
+  }
+}
+
+Status GcsClientImpl::Connect() {
```
Comment (raulchen) [inline src/ray/gcs/gcs_client_impl.cc]: we should check `is_connected` first.

File: src/ray/gcs/call_back.h (hunk 10 above)
Comment (micafan) [reply src/ray/gcs/call_back.h]: done

File: src/ray/gcs/actor_state_accessor.cc (hunk 8 above)
Comment (micafan) [reply src/ray/gcs/actor_state_accessor.cc]: on_success

**Pull request closed without merging**